### DISCUSSION TITLE
Internal note links semantics

### DISCUSSION THREAD

**Discussion by peter-facko** [TriliumNext/Trilium#3831] [General]
Hi,

Trilium has the feature of linking notes in a note's text. When creating the link and searching by name and also when hovering over the link, *a* path to the note is shown. However, when the destination note is cloned in multiple places, only one path is shown in both cases. I don't know by what logic the clone and therefore the shown path is chosen.

This opens a question whether the link should be semantically thought of as storing a path or a reference to a note. The behavior is that of note reference. But searching for the note or hovering over the link shows just one path, which is in my opinion confusing, because it hints that it is somehow also referring to the specific clone.

This also leads to a question whether it isn't often the case that a note has multiple clones but one is thought of by the user as the "main" or "owning" and others as references. I think the analogy from programming is accurate; a reference acts the same as the owning variable when used, but it is itself a different type. I haven't yet thought of a purpose for this distinction other than the path shown when creating an internal link, but it seems to me that it might become important in the future or already is hidden in some problem in the app.

I don't have any specific question, just feel free to comment your thoughts on the above.

**Comment by zadam** [comment]
> However, when the destination note is cloned in multiple places, only one path is shown in both cases. I don't know by what logic the clone and therefore the shown path is chosen.

That's true, and this is intentional, because in most cases people don't care that much which clone is used and showing all clones could lead to "explosion" of results which might users consider to be useless duplicates filling in too much space.

> This opens a question whether the link should be semantically thought of as storing a path or a reference to a note.

It is meant to be storing a path. But because paths can change (e.g. when moving the note, the original links in text notes are not updated to reflect the new path), there's a fallback that at worst it's just a reference to a note. If the exact note path does not exist, there's some smartness in trying to find the closest path.

> This also leads to a question whether it isn't often the case that a note has multiple clones but one is thought of by the user as the "main" or "owning" and others as references.

Yes, I think many people often think of it this way and the name "clone" is also suggesting this (it's a bad name, but I still couldn't figure out a better one). Technically / semantically, there's no "main" or "owning" clone, all are equal.

I would like to point out that when e.g. searching / linking, it's not an arbitrary clone which is being displayed, but a clone which best matches the search string. Consider this setup:

<img width="341" alt="image" src="https://user-images.githubusercontent.com/617641/232117121-e8144e7b-8712-4f37-a6bf-546dc86942a4.png">

Now linking:

<img width="530" alt="image" src="https://user-images.githubusercontent.com/617641/232117270-119363b3-1857-4c54-bcc5-1210ed0b3c1d.png">

Multiple clones match that search string, we want to display just one, so we pick the closest one to the root (considered to be the most important).

Now a different search string:

<img width="542" alt="image" src="https://user-images.githubusercontent.com/617641/232117519-d7d1018b-edf0-475a-9f49-2d6330355066.png">

Only the second clone matches the search string (because the note path is taken into account) and that one will be linked.
